kani-compiler/src/codegen_cprover_gotoc/reachability.rs

```diff
@@ -0,0 +1,586 @@
+use rustc_middle::mir::mono::MonoItem;
+use rustc_middle::mir::visit::Visitor as MirVisitor;
+use rustc_middle::mir::{
+    Body, CastKind, Constant, ConstantKind, Location, Rvalue, Terminator, TerminatorKind,
+};
+use rustc_middle::span_bug;
+use rustc_middle::ty::adjustment::PointerCast;
+use rustc_middle::ty::layout::{
+    HasParamEnv, HasTyCtxt, LayoutError, LayoutOf, LayoutOfHelpers, TyAndLayout,
+};
+use rustc_middle::ty::{
+    Closure, ClosureKind, Const, ConstKind, Instance, InstanceDef, ParamEnv, Ty, TyCtxt, TyKind,
+    TypeAndMut, TypeFoldable, VtblEntry,
+};
+use rustc_span::def_id::DefId;
+use rustc_span::source_map::Span;
+use rustc_target::abi::{HasDataLayout, TargetDataLayout, VariantIdx};
+use tracing::{debug, debug_span, trace, warn};
+
+struct MonoItemsCollector<'tcx> {
+    /// The compiler context.
+    tcx: TyCtxt<'tcx>,
+    /// Set of collected items used to avoid entering recursion loops.
+    collected: FxHashSet<MonoItem<'tcx>>,
+    /// Keep items collected sorted by visiting order which should be more stable.
+    collected_sorted: Vec<MonoItem<'tcx>>,
+    /// Items enqueued for visiting.
+    queue: Vec<MonoItem<'tcx>>,
+}
+
+impl<'tcx> MonoItemsCollector<'tcx> {
+    /// Collects all reachable items starting from the given root.
+    pub fn collect(&mut self, root: MonoItem<'tcx>) {
+        debug!(?root, "collect");
+        self.queue.push(root);
+        self.reachable_items();
+    }
+
+    /// Traverses the call graph starting from the given root. For every function, we visit all
+    /// instruction looking for the items that should be included in the compilation.
+    fn reachable_items(&mut self) {
+        while !self.queue.is_empty() {
+            let to_visit = self.queue.pop().unwrap();
+            if !self.collected.contains(&to_visit) {
+                self.collected.insert(to_visit);
+                self.collected_sorted.push(to_visit);
+                match to_visit {
+                    MonoItem::Fn(instance) => {
+                        self.visit_fn(instance);
+                    }
+                    MonoItem::Static(def_id) => {
+                        self.visit_static(def_id);
+                    }
+                    MonoItem::GlobalAsm(_) => {
+                        self.visit_asm(to_visit);
+                    }
+                }
+            }
+        }
+    }
+
+    /// Visit a function and collect all mono-items reachable from its instructions.
+    fn visit_fn(&mut self, instance: Instance<'tcx>) {
+        let _guard = debug_span!("visit_fn", function=?instance).entered();
+        let body = self.tcx.instance_mir(instance.def);
+        let mut collector =
+            MonoItemsFnCollector { tcx: self.tcx, collected: FxHashSet::default(), instance, body };
+        collector.visit_body(body);
+        self.queue.extend(collector.collected.iter().filter(|item| !self.collected.contains(item)));
```

Comment:
1. Indeed, I'll either change collector.collected to be a vector or use BTreeMap.
2. I can probably just remove the filter since we already check if the item has been visited in the main loop. This was just to keep the queue small.

Reply:
I fixed this by sorting everything at the end. It's cleaner and less error prone. My main concern is performance, but I don't have any good way to measure that today.